File: HW/HW9/hw9.py
def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
    if not root:
        return None
    
    root_val = root.val
    p_val = p.val
    q_val = q.val
    
    if p_val < -1000000000 or p_val > 1000000000 or q_val < -1000000000 or q_val > 1000000000:
        raise ValueError("Node value out of bounds")
    
    if p_val == q_val:
        raise ValueError("p and q should be distinct nodes")
        
    if p_val < root_val and q_val < root_val:
        return self.lowestCommonAncestor(root.left, p, q)
    elif p_val > root_val and q_val > root_val:
        return self.lowestCommonAncestor(root.right, p, q)
    else:
        return root

File: HW/HW9/hw9.java
public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    if (root == null) {
        return null;
    }
    
    int rootVal = root.val;
    int pVal = p.val;
    int qVal = q.val;
    
    if (pVal < -1000000000 || pVal > 1000000000 || qVal < -1000000000 || qVal > 1000000000) {
        throw new IllegalArgumentException("Node value out of bounds");
    }
    
    if (pVal == qVal) {
        throw new IllegalArgumentException("p and q should be distinct nodes");
    }
    
    if (pVal < rootVal && qVal < rootVal) {
        return lowestCommonAncestor(root.left, p, q);
    } else if (pVal > rootVal && qVal > rootVal) {
        return lowestCommonAncestor(root.right, p, q);
    } else {
        return root;
    }
}
